Prototype for building non-strict leak detector

tugsbayasgalan opened the pull request:
Summary:
Our strategy for detecting fake tensor leakage in non-strict for outside scope (side effects happening outside of model.forward) is:
1. We do gc.collect() before export and get the alive fake tensors 
2. We dump the proxy to fake tensor map from make_fx tracer
3. We query gc again to get alive fake tensors 
4. We take the delta between (1) and (3)
5. Filter out fake tensors that are:
    1. Associated with `TrackedFake` (input tracking thing in symbolic_shapes)
    2. Associated with `gm.meta` 
6. Do ID match with the proxies and emit their stacktraces. 


We rely on (https://github.com/pytorch/pytorch/pull/159923) for other sources of leakages such as:
1. We failed to proxy an operator (like param.data) 
2. We cache some tensor in model.forward (https://github.com/pytorch/pytorch/issues/155114)


In general, we notice `gc.collect()` and query-ing gc for live objects are kinda slow. So we turn on this feature under env variable. We should document on export public facing documents that if you run into weird errors regarding fake tensors, they should look into turning on this env variable for further analysis.

Test Plan:
Test plan

Rollback Plan:

Differential Revision: D80003204




cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv

avikchaudhuri commented:
Is the goal here to detect fake tensor leakage into arbitrary global state or under the exported model itself?

If the former, I'm fine with this approach. Is the latter already covered with detect attribute assignment?

tugsbayasgalan commented:
> Is the goal here to detect fake tensor leakage into arbitrary global state or under the exported model itself?
> 
> If the former, I'm fine with this approach. Is the latter already covered with detect attribute assignment?

It is former. The latter is covered by "detect attribute assignment" PR: https://github.com/pytorch/pytorch/pull/159923. In general we can think of this goal as several mini components.

tugsbayasgalan commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -1866,6 +1869,25 @@ def trace(  # type: ignore[override]
     ) -> fx.Graph:
         res = super().trace(root, concrete_args)
 
+        # NOTE [export non-strict fake tensor leak detection]
```
This is the module stack tracer, a non-strict version of proxy tracer where we populate nn_module_stack.

tugsbayasgalan commented on torch/export/_leakage_detection_utils.py:
```diff
@@ -0,0 +1,89 @@
+    types.MethodType,
+    types.CodeType,
+    types.ModuleType,
+)
```
Actually we don't want to skip some of it. Removed

tugsbayasgalan committed:
Prototype for building non-strict leak detector (#160456)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/160456

Our strategy for detecting fake tensor leakage in non-strict for outside scope (side effects happening outside of model.forward) is:
(1) We do gc.collect() before export and get the alive fake tensors
(2) We dump the proxy to fake tensor map from make_fx tracer
(3) We query gc again to get alive fake tensors
(4) We take the delta between (1) and (3)
(5) Filter out fake tensors that are:
Associated with TrackedFake (input tracking thing in symbolic_shapes)
Associated with gm.meta
(6) Do ID match with the proxies

In general, we notice gc.collect() and query-ing gc for live objects are kinda slow. So we turn on this feature under env variable. We should document on export public facing documents that if you run into weird errors regarding fake tensors, they should look into turning on this env variable for further analysis.

Test Plan:
Test plan

Rollback Plan:

Differential Revision: D80003204
